aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.s3;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.relocated.com.google.common.hash.HashCode;
+import org.apache.iceberg.relocated.com.google.common.hash.HashFunction;
+import org.apache.iceberg.relocated.com.google.common.hash.Hashing;
+import org.apache.iceberg.util.LocationUtil;
+
+/**
+ * This location provider provides data locations that are optimized for S3 performance. Both
+ * General Purpose buckets and Directory buckets will see better throughput and autoscaling behavior
+ * than using the generic ObjectStoreLocationProvider.
+ *
+ * <p>The data location is resolved as follows. Data files are written directly at this path with no
+ * other intermediate directories created.
+ *
+ * <ol>
+ *   <li>{@link TableProperties#WRITE_DATA_LOCATION}
+ *   <li><code>tableLocation + "/data"</code>
+ * </ol>
+ *
+ * The data file is placed immediately under the data location. Partition names are <b>not</b>
```

Comment:
I'm concerned about not including an option for partition values.  I generally agree that we don't need them and it's better for the overall key size to exclude that info, but I feel like it should still be an option and shouldn't impact the partition distribution.

Reply:
Intention of this change is to not only provide an optimal performance for S3 General Purpose Buckets but also for S3 Directory Buckets. This change essentially minimizes the directories required in the Directory Buckets, which have directories as a first-class citizen, whereas General Purpose buckets have a flat key space.
 
S3 General Purpose Buckets performance best practices suggest using entropy in prefixes to distribute load which the base2 achieves here but since Directory Buckets internally manage load distribution, creating sparse directories with that level of entropy can cause requests to be slower. So we wanted to provide a single entry point for all S3 relevant optimizations that captures these 2 bucket types.

Reply:
@ookumuso it's helpful to understand the context behind why you may not want include the partition structure in the path, but I still think it should be an option (even if suboptimal).

Reply:
Sounds like this can be just configurable, with a property like `write.location-provider.omit-partition-values-in-path`? We have seen people depending business workflows on partition path values which should be strongly discouraged, so I think it is better to default to not having partition structure in path. @danielcweeks what do you think?